src/backend/d3d12/PipelineLayoutD3D12.cpp

```diff
@@ -24,10 +24,117 @@ namespace d3d12 {
+        union {
+            D3D12_ROOT_DESCRIPTOR_TABLE DescriptorTable;
+            D3D12_ROOT_CONSTANTS Constants;
+            D3D12_ROOT_DESCRIPTOR Descriptor;
+        } rootParameterValues[kMaxBindGroups * 2];
+        // samplers must be in a separate descriptor table so we need at most twice as many tables as bind groups
+
+        // Ranges are D3D12_DESCRIPTOR_RANGE_TYPE_(SRV|UAV|CBV|SAMPLER)
+        // They are grouped together so each table as at most 4 ranges
+        D3D12_DESCRIPTOR_RANGE ranges[kMaxBindGroups * 4];
+
+        uint32_t parameterIndex = 0;
+        uint32_t rangeIndex = 0;
+
+        for (uint32_t group = 0; group < kMaxBindGroups; ++group) {
+            const auto& groupInfo = GetBindGroupLayout(group)->GetBindingInfo();
+
+            descriptorCountInfo[group] = {};
+            uint32_t& CBVGroupIndex = descriptorCountInfo[group][ToIndex(Descriptor::Type::CBV)];
+            uint32_t& UAVGroupIndex = descriptorCountInfo[group][ToIndex(Descriptor::Type::UAV)];
+            uint32_t& SRVGroupIndex = descriptorCountInfo[group][ToIndex(Descriptor::Type::SRV)];
+            uint32_t& SamplerGroupIndex = descriptorCountInfo[group][ToIndex(Descriptor::Type::Sampler)];
+
+            for (uint32_t binding = 0; binding < kMaxBindingsPerGroup; ++binding) {
+                if (!groupInfo.mask[binding]) {
+                    continue;
+                }
+
+                switch (groupInfo.types[binding]) {
+                    case nxt::BindingType::UniformBuffer:
+                        CBVGroupIndex++;
+                        break;
+                    case nxt::BindingType::StorageBuffer:
+                        UAVGroupIndex++;
+                        break;
+                    case nxt::BindingType::Sampler:
+                        SRVGroupIndex++;
+                        break;
+                    case nxt::BindingType::SampledTexture:
+                        SamplerGroupIndex++;
+                        break;
+                }
+            }
+
+            if (CBVGroupIndex > 0 || UAVGroupIndex > 0 || SRVGroupIndex > 0) {
+                auto& rootParameter = rootParameters[parameterIndex];
+                rootParameter.ParameterType = D3D12_ROOT_PARAMETER_TYPE_DESCRIPTOR_TABLE;
+                rootParameter.DescriptorTable = rootParameterValues[parameterIndex].DescriptorTable;
+                rootParameter.ShaderVisibility = D3D12_SHADER_VISIBILITY_ALL;
+                rootParameter.DescriptorTable.NumDescriptorRanges = 0;
+                rootParameter.DescriptorTable.pDescriptorRanges = &ranges[rangeIndex];
+
+                if (CBVGroupIndex > 0) {
+                    rootParameter.DescriptorTable.NumDescriptorRanges++;
+                    auto& range = ranges[rangeIndex++];
+                    range.RangeType = D3D12_DESCRIPTOR_RANGE_TYPE_CBV;
+                    range.NumDescriptors = CBVGroupIndex;
+                    range.BaseShaderRegister = group * kMaxBindingsPerGroup;
```

Comment:
Done. They're computed in BindGroupLayout and the ranges are copied over and then the BaseShaderRegister is set.